homework-5/homework-5.py

```diff
@@ -0,0 +1,86 @@
+# Backend Programming with Python
+# Homework 5 - Battleships
+
+import random, math
+
+
+class SpaceShip:
+    def __init__(self, name, hp, armour, damage):
+        self.name = name
+        self.hp = hp
+        self.armour = armour
+        self.damage = damage
+
+    def attack(self, other):
+        if isinstance(other, SpaceShip):
+            damage_dealt = self.damage - other.armour
+            if damage_dealt < 0 < self.damage:
+                damage_dealt = 1
+            other.hp = other.hp - damage_dealt
+        print(
+            f'Damage dealt to {other.name} = {self.damage} ({self.name} Damage) - {other.armour} ({other.name} Armor) = {damage_dealt}')
+        print(f'HP left on {other.name} = {other.hp}')
+
+    def upgrade(self, upgrade_name):
+        if upgrade_name == 'Titanium Armour':
+            self.armour = self.armour + 3
+            self.hp = self.hp + 250
+        elif upgrade_name == 'Absorption Shield':
+            self.armour = math.floor(self.armour * 1.5)
+        elif upgrade_name == 'Proton Torpedos':
+            self.damage = self.damage + 25
+        elif upgrade_name == 'Flare Engine':
+            self.damage = math.floor(self.damage * 1.5)
+            self.armour = self.armour + 3
+            self.hp = math.floor(self.hp * 1.2)
+
+
+def main():
+    ship1 = SpaceShip('Nostromo', random.randint(1, 100), random.randint(0, 30), random.randint(0, 30))
+    ship2 = SpaceShip('Covenant', random.randint(1, 100), random.randint(0, 30), random.randint(0, 30))
+
+    print(f'Stats before upgrades:\n'
+          f'Ship 1: Name = {ship1.name}, HP = {ship1.hp}, Armour = {ship1.armour}, Damage = {ship1.damage}\n'
+          f'Ship 2: Name = {ship2.name}, HP = {ship2.hp}, Armour = {ship2.armour}, Damage = {ship2.damage}')
+
+    ship1_upgrades = list()
+    ship2_upgrades = list()
+    while True:
+        a = input('Do you want to upgrade a ship? (y/n)  ')
+        if a == 'y':
+            ship_name = input("Enter ship's name:\nOptions:\nNostromo or Covenant:  ")
+            upgrade_name = input("Now enter the upgrade's name:\n"
+                                 "Options:\nTitanium Armour, Absorption Shield, Proton Torpedos, and Flare Engine: ")
+            if ship_name == ship1.name and upgrade_name not in ship1_upgrades:
```

Comment:
The code duplication here between the two branches suggests to me you are not very comfortable with playing with instances and variables. Here is my recommendation for improvement:
* First determine the ship that would be affected, validate input, and put its reference in a separate variable
* Next determine the upgrade that would be added, validate input, store it in a variable
* Last just add the upgrade to the ship.